### PR TITLE
Re-enable extensibility points for ExpressionHelper.GetExpressionText, ExpressionMetadataProvider.FromStringExpression

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes https://github.com/aspnet/Mvc/issues/8724

### CODE REVIEW

**Comment by pranavkm** [comment]
* Makes `HtmlHelper<T>.GetExpressionName` public instead of making `ExpressionHelper` public.
* `IModelExpressionProvider.CreateModelExpression` - https://github.com/aspnet/Mvc/blob/master/src/Microsoft.AspNetCore.Mvc.ViewFeatures/ModelExpressionProvider.cs#L42-L65 - could be used in place of `ExpressionMetadataProvider.FromLambdaExpression`. The change was to add another overload to support the `ExpressionMetadataProvider.FromStringExpression` equivalent.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ExpressionHelper.cs [hunk 1]
```diff
@@ -11,28 +12,32 @@
 
 namespace Microsoft.AspNetCore.Mvc.ViewFeatures
 {
-    internal static class ExpressionHelper
+    internal class ExpressionHelper
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ExpressionHelper.cs]: We use this as a singleton service and it takes care of the caching.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ExpressionTextCache.cs [hunk 2]
```diff
@@ -1,124 +0,0 @@
...
-using System.Collections.Generic;
-using System.Linq.Expressions;
-using Microsoft.Extensions.Internal;
-
-namespace Microsoft.AspNetCore.Mvc.ViewFeatures
-{
-    /// <summary>
-    /// This class holds the cache for the expression text that is computed by ExpressionHelper.
-    /// </summary>
-    public class ExpressionTextCache
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ExpressionTextCache.cs]: Extracted `LambdaExpressionComparer` into a separate `internal` type as-is. Git doesn't like the diff so it's treating this as a add-delete

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelper.cs [hunk 3]
```diff
@@ -1246,5 +1246,11 @@ protected virtual IEnumerable<SelectListItem> GetEnumSelectList(ModelMetadata me
 
             return selectList;
         }
+
+        private static string GetExpressionText(string expression)
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelper.cs]: Moved this from `ExpressionHelper` It's pretty trivial

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs [hunk 4]
```diff
@@ -89,6 +80,8 @@ public override void Contextualize(ViewContext viewContext)
                     nameof(viewContext));
             }
 
+            _expressionHelper = viewContext.HttpContext.RequestServices.GetRequiredService<ExpressionHelper>();
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs]: Kinda iffy, but it's the least breaking option. Another alternative is to make the ctor `internal` and use a "factory" method in DI to new this up. I'm not sure it's worth the trouble.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs [hunk 5]
```diff
@@ -370,14 +363,20 @@ public string NameFor<TResult>(Expression<Func<TModel, TResult>> expression)
...
             if (expression == null)
             {
                 throw new ArgumentNullException(nameof(expression));
             }
 
-            return ExpressionHelper.GetExpressionText(expression, _expressionTextCache);
+            return _expressionHelper.GetExpressionText(expression);
         }
 
         protected ModelExplorer GetModelExplorer<TResult>(Expression<Func<TModel, TResult>> expression)
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs]: I could make this `public` too. Would help in the `extension` scenarios without having to get `IModelExpressionProvider` from DI. Thoughts?

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/LambdaExpressionComparer.cs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq.Expressions;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures
+{
+    // This comparer is tightly coupled with the logic of ExpressionHelper.GetExpressionText.
+    // It is not designed to accurately compare any two arbitrary LambdaExpressions.
+    internal class LambdaExpressionComparer : IEqualityComparer<LambdaExpression>
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/LambdaExpressionComparer.cs]: Moved as is

**Comment by pranavkm** [comment]
cc @San4es

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs [hunk 7]
```diff
@@ -370,14 +363,20 @@ public string NameFor<TResult>(Expression<Func<TModel, TResult>> expression)
...
         }
 
-        protected string GetExpressionName<TResult>(Expression<Func<TModel, TResult>> expression)
+        /// <summary>
+        /// Gets the name for the lambda expression.
+        /// </summary>
+        /// <typeparam name="TResult">The type of the result.</typeparam>
+        /// <param name="expression">The expression.</param>
+        /// <returns>The expression name.</returns>
+        public string GetExpressionName<TResult>(Expression<Func<TModel, TResult>> expression)
```
Comment (San4es) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs]: @pranavkm It is replacement for `ExpressionHelper.GetExpressionText(expression)`, isn't it?

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ModelExpressionProvider.cs [hunk 8]
```diff
@@ -63,5 +63,30 @@ public class ModelExpressionProvider : IModelExpressionProvider
 
             return new ModelExpression(name, modelExplorer);
         }
+
+        /// <inheritdoc />
+        public ModelExpression CreateModelExpression<TModel>(
```
Comment (San4es) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ModelExpressionProvider.cs]: @pranavkm Is it a replacement for ExpressionMetadataProvider.FromStringExpression()? If so, how can I access to it while `ModelExpressionProvider` is internal?

**Review by San4es — COMMENTED** [review]
@pranavkm And what about `ExpressionMetadataProvider.FromLambdaExpression`?

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs (hunk 7 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs]: Yup.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ModelExpressionProvider.cs (hunk 8 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.ViewFeatures/ModelExpressionProvider.cs]: The interface `IModelExpressionProvider` is a wrapper for `ExpressionMetadataProvider`. It returns an instance `ModelExplorer` which lets you get to the `ModelMetadata` of the type and the graph of it's properties.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs [hunk 9]
```diff
@@ -10,24 +10,21 @@
...
 namespace Microsoft.AspNetCore.Mvc.ViewFeatures
 {
     public class HtmlHelper<TModel> : HtmlHelper, IHtmlHelper<TModel>
     {
-        private readonly ExpressionTextCache _expressionTextCache;
+        private ExpressionHelper _expressionHelper;
 
-        /// <summary>
-        /// Initializes a new instance of the <see cref="HtmlHelper{TModel}"/> class.
-        /// </summary>
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs]: This is a public class (though I wish it weren't). Don't remove doc comments.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs (hunk 5 above)
Comment (dougbu) [reply src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs]: I think `public` is good here but both changes require a down-cast from `IHtmlHelper<T>` and that's not great. Could we add an additional `interface` and register this class using that too?

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/IModelExpressionProvider.cs [hunk 10]
```diff
@@ -23,5 +23,17 @@ public interface IModelExpressionProvider
...
+
+        /// <summary>
+        /// Returns a <see cref="ModelExpression"/> instance describing the given <paramref name="expression"/>.
+        /// </summary>
+        /// <typeparam name="TModel">The type of the <paramref name="viewData"/>'s <see cref="ViewDataDictionary{T}.Model"/>.</typeparam>
+        /// <param name="viewData">The <see cref="ViewDataDictionary{TModel}"/> containing the <see cref="ViewDataDictionary{T}.Model"/> 
+        /// against which <paramref name="expression"/> is evaluated. </param>
+        /// <param name="expression">Expression name, relative to <c>viewData.Model</c>.</param>
+        /// <returns>A new <see cref="ModelExpression"/> instance describing the given <paramref name="expression"/>.</returns>
+        ModelExpression CreateModelExpression<TModel>(
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/IModelExpressionProvider.cs]: Why isn't this a breaking change?

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs [hunk 11]
```diff
@@ -10,24 +10,21 @@
...
         public HtmlHelper(
             IHtmlGenerator htmlGenerator,
             ICompositeViewEngine viewEngine,
             IModelMetadataProvider metadataProvider,
             IViewBufferScope bufferScope,
             HtmlEncoder htmlEncoder,
-            UrlEncoder urlEncoder,
-            ExpressionTextCache expressionTextCache)
+            UrlEncoder urlEncoder)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs]: Isn't this change to a `public` ctor breaking?

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs (hunk 4 above)
Comment (dougbu) [reply src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs]: Well, if we could break things that much, would be great to make `DefaultHtmlGenerator`, `HtmlHelper` and `HtmlHelper<T>` `internal`. Fall-back would be to make `DefaultHtmlGenerator` and `HtmlHelper<T>` `sealed` and remove any `protected` or `virtual` members not used in test derived classes (if any).

**Review by dougbu — CHANGES_REQUESTED** [review]
Might be fine but&hellip;

**Review by pranavkm — COMMENTED** [review]
🆙 📅

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs [hunk 12]
```diff
@@ -209,10 +197,10 @@ public string DisplayTextFor<TResult>(Expression<Func<TModel, TResult>> expressi
                 throw new ArgumentNullException(nameof(expression));
             }
 
-            var modelExplorer = GetModelExplorer(expression);
+            var modelExpression = GetModelExpression(expression);
             return GenerateEditor(
-                modelExplorer,
-                htmlFieldName ?? GetExpressionName(expression),
+                modelExpression.ModelExplorer,
+                htmlFieldName ?? modelExpression.Name,
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs]: This seems to be the only place where there's an additional `GetExpressionName` call where previously there was none. Using `GetModelExpression` seems more consistent than optimizing for the non-null case

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs [hunk 13]
```diff
@@ -361,41 +349,62 @@ public string NameFor<TResult>(Expression<Func<TModel, TResult>> expression)
...
                 htmlAttributes);
         }
 
+        /// <summary>
+        /// Gets the <see cref="ModelExpression"/> for <paramref name="expression"/>.
+        /// </summary>
+        /// <typeparam name="TResult">The type of the result.</typeparam>
+        /// <param name="expression">The expression.</param>
+        /// <returns>The <see cref="ModelExpression"/>.</returns>
+        public ModelExpression GetModelExpression<TResult>(Expression<Func<TModel, TResult>> expression)
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs]: I'm neutral about making them public - it does let users create `HtmlHelper<T>` extensions or downcast and use this methods without introducing additional test burden

**Comment by pranavkm** [comment]
Closing in favor of https://github.com/aspnet/AspNetCore/pull/4494/files

**Pull request closed without merging**